### ISSUE TITLE
[Microsoft365] Can't connect any longer 'BAD User is authenticated but not connected'

### ISSUE THREAD

**Issue by DierkDroth** [jstedfast/MailKit#1494]
@jstedfast unfortunately I suddenly experience more issues with MS365: I no longer can connect but get the error `BAD User is authenticated but not connected`
(details below)

I'm pretty sure I have not changed my connect logic. So, I would not know what might be causing the issue.
This is latest MailKit 3.4.3 on .NET 7.0 + Windows 11.

Would you have any idea?

```
2022-12-27 07:41:00:283 Imap.Connected: 'imaps://outlook.office365.com:993/'
2022-12-27 07:41:00:353 Imap.S: * OK The Microsoft Exchange IMAP4 service is ready. [RgBSADMAUAAyADgAMQBDAEEAMAAxADUANwAuAEQARQBVAFAAMgA4ADEALgBQAFIATwBEAC4ATwBVAFQATABPAE8ASwAuAEMATwBNAA==]
2022-12-27 07:41:00:363 Imap.C: A00000000 CAPABILITY
2022-12-27 07:41:00:403 Imap.S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
2022-12-27 07:41:00:403 Imap.S: A00000000 OK CAPABILITY completed.
2022-12-27 07:41:00:700 Imap.C: A00000001 AUTHENTICATE XOAUTH2 <some token>=
2022-12-27 07:41:01:763 Imap.S: A00000001 OK AUTHENTICATE completed.
2022-12-27 07:41:01:764 Imap.C: A00000002 CAPABILITY
2022-12-27 07:41:01:811 Imap.S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
2022-12-27 07:41:01:811 Imap.S: A00000002 OK CAPABILITY completed.
2022-12-27 07:41:01:813 Imap.C: A00000003 NAMESPACE
2022-12-27 07:41:01:859 Imap.S: A00000003 BAD User is authenticated but not connected.
```

**Comment by DierkDroth** [comment]
@jstedfast this is absolutely weird: I'm travelling these days and connecting to the internet through my 4G/5G iPhone hotspot.

I can reproduce the issue above 100% when to my iPhone's hotspot. However, as I- the very same minute - switched to the hotel's WiFi, the issue is gone.

How would MS's infrastructure toggle behavior if I'm connected through iPhone hotspot or hotel's WiFi?!? I'm baffled...

Here is the 'OK' log for reference:
```
2022-12-28 15:10:52:990 Imap.S: * OK The Microsoft Exchange IMAP4 service is ready. [RgBSADIAUAAyADgAMQBDAEEAMAAxADcAMgAuAEQARQBVAFAAMgA4ADEALgBQAFIATwBEAC4ATwBVAFQATABPAE8ASwAuAEMATwBNAA==]
2022-12-28 15:10:52:999 Imap.C: A00000000 CAPABILITY
2022-12-28 15:10:53:020 Imap.S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
2022-12-28 15:10:53:020 Imap.S: A00000000 OK CAPABILITY completed.
2022-12-28 15:10:53:284 Imap.C: A00000001 AUTHENTICATE XOAUTH2 <token>=
2022-12-28 15:10:53:908 Imap.S: A00000001 OK AUTHENTICATE completed.
2022-12-28 15:10:53:909 Imap.C: A00000002 CAPABILITY
2022-12-28 15:10:53:933 Imap.S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
2022-12-28 15:10:53:933 Imap.S: A00000002 OK CAPABILITY completed.
2022-12-28 15:10:53:935 Imap.C: A00000003 NAMESPACE
2022-12-28 15:10:53:956 Imap.S: * NAMESPACE (("" "/")) NIL NIL
2022-12-28 15:10:53:956 Imap.S: A00000003 OK NAMESPACE completed.
2022-12-28 15:10:53:960 Imap.C: A00000004 LIST "" "INBOX"
2022-12-28 15:10:53:998 Imap.S: * LIST (\Marked \HasChildren) "/" INBOX
2022-12-28 15:10:53:998 Imap.S: A00000004 OK LIST completed.
2022-12-28 15:10:54:003 Imap.C: A00000005 SELECT INBOX
2022-12-28 15:10:54:058 Imap.S: * 3 EXISTS
2022-12-28 15:10:54:058 Imap.S: * 0 RECENT
2022-12-28 15:10:54:058 Imap.S: * FLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)
2022-12-28 15:10:54:058 Imap.S: * OK [PERMANENTFLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)] Permanent flags
2022-12-28 15:10:54:058 Imap.S: * OK [UIDVALIDITY 14] UIDVALIDITY value
2022-12-28 15:10:54:058 Imap.S: * OK [UIDNEXT 9026] The next unique identifier value
2022-12-28 15:10:54:058 Imap.S: A00000005 OK [READ-WRITE] SELECT completed.
2022-12-28 15:10:54:067 Imap.C: A00000006 LIST "" "%"
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Archive
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasChildren) "/" Calendar
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasChildren) "/" Contacts
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasChildren) "/" "Conversation History"
2022-12-28 15:10:54:101 Imap.S: * LIST (\Marked \HasNoChildren \Trash) "/" "Deleted Items"
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Done
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren \Drafts) "/" Drafts
2022-12-28 15:10:54:101 Imap.S: * LIST (\Marked \HasChildren) "/" INBOX
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Journal
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren \Junk) "/" "Junk Email"
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Notes
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Outbox
2022-12-28 15:10:54:101 Imap.S: * LIST (\Marked \HasNoChildren \Sent) "/" "Sent Items"
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasChildren) "/" "Sync Issues"
2022-12-28 15:10:54:101 Imap.S: * LIST (\HasNoChildren) "/" Tasks
2022-12-28 15:10:54:101 Imap.S: A00000006 OK LIST completed.
2022-12-28 15:10:54:103 Imap.C: A00000007 SUBSCRIBE Done
2022-12-28 15:10:54:127 Imap.S: A00000007 OK SUBSCRIBE completed.
2022-12-28 15:10:54:127 Imap.C: A00000008 LIST "" "%"
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Archive
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasChildren) "/" Calendar
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasChildren) "/" Contacts
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasChildren) "/" "Conversation History"
2022-12-28 15:10:54:157 Imap.S: * LIST (\Marked \HasNoChildren \Trash) "/" "Deleted Items"
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Done
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren \Drafts) "/" Drafts
2022-12-28 15:10:54:157 Imap.S: * LIST (\Marked \HasChildren) "/" INBOX
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Journal
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren \Junk) "/" "Junk Email"
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Notes
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Outbox
2022-12-28 15:10:54:157 Imap.S: * LIST (\Marked \HasNoChildren \Sent) "/" "Sent Items"
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasChildren) "/" "Sync Issues"
2022-12-28 15:10:54:157 Imap.S: * LIST (\HasNoChildren) "/" Tasks
2022-12-28 15:10:54:157 Imap.S: A00000008 OK LIST completed.
2022-12-28 15:10:54:157 Imap.C: A00000009 LIST "" "%"
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Archive
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasChildren) "/" Calendar
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasChildren) "/" Contacts
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasChildren) "/" "Conversation History"
2022-12-28 15:10:54:188 Imap.S: * LIST (\Marked \HasNoChildren \Trash) "/" "Deleted Items"
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Done
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren \Drafts) "/" Drafts
2022-12-28 15:10:54:188 Imap.S: * LIST (\Marked \HasChildren) "/" INBOX
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Journal
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren \Junk) "/" "Junk Email"
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Notes
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Outbox
2022-12-28 15:10:54:188 Imap.S: * LIST (\Marked \HasNoChildren \Sent) "/" "Sent Items"
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasChildren) "/" "Sync Issues"
2022-12-28 15:10:54:188 Imap.S: * LIST (\HasNoChildren) "/" Tasks
2022-12-28 15:10:54:188 Imap.S: A00000009 OK LIST completed.

```

**Comment by jstedfast** [comment]
I don't know. The only scenario that I've seen this happen for is when the user doesn't have the proper IMAP permissions configured for their account.

I would recommend searching for this error on https://stackoverflow.com - most likely someone else has hit this with other libraries as well and maybe someone has figured out a solution.

**Comment by DierkDroth** [comment]
Thanks @jstedfast.

FYI I also have a Google account authorized through OAuth2 -> no issue. Same thing for a few other mail accounts connected by legacy user/pwd -> no issue. It's really just the MS 365 (OAuth2) account which shows this weird behavior...

**Comment by ekalchev** [comment]
We observed the same issue. We contacted Microsoft support, but they couldn't tell us why this happens. I believe this is somehow related to the amount of requests sent within a certain time window. The Exchange server (which is used for IMAP) has limits, and if you exceed any of them, the server will refuse to serve the requests you sent. I believe the behavior we are observing is the same, but without a meaningful error message. My advice is to try to reduce the number of IMAP commands sent and see if this issue disappears. We resolved the problem in our mail app by implementing the Exchange protocol. I think Microsoft intentionally has a poor implementation of IMAP. There are also other issues with their IMAP, such as notifications for new emails taking too long.

**Comment by DierkDroth** [comment]
@ekalchev thanks, that is an interesting insight!

> My advice is to try to reduce the number of IMAP commands sent 

Would you mind sharing some thoughts on how that could be done based on the logs in my first post? I suspect @jstedfast's implementation only would apply those IMAP commands which are absolutely required. Hence, there might be no/limited room for improvements ... just guessing

**Comment by jstedfast** [comment]
Mostly what you want to do is make sure that you are batching requests.

For example:

```csharp
var summaries = folder.Fetch (0, -1, ...);
foreach (var summary in summaries) {
    var message = folder.GetMessage (summary.UniqueId);
    folder.AddFlags (summary.UniqueId, MessageFlags.Deleted, true);
}
```

Would be more efficient as:

```csharp
var summaries = folder.Fetch (0, -1, ...);
var uids = new UniqueIdSet ();
foreach (var summary in summaries) {
    var message = folder.GetMessage (summary.UniqueId);
    uids.Add (summary.UniqueId);
}

folder.AddFlags (uids, MessageFlags.Deleted, true);
```

If `summaries` contains 1000 items, then the original snippet would send 1000 commands to set the `\Deleted` flag while the second snippet would mark all of those same messages for deletion in a single command.

**Comment by DierkDroth** [comment]
Thanks @jstedfast

To clarify: this is a general recommendation or do you actually see something in my logs which I could improve?

**Comment by jstedfast** [comment]
It's just a general rule. I don't see enough in your logs to offer any specific advice.

**Comment by ekalchev** [comment]
We switched to Exchange protocol for O365 servers because reducing the imap commands cannot be guaranteed for every account. For example if you want to sync flags for an inbox of 20k emails, the server will reply with 20k rows for each email. I found that we are hitting this issue with microsoft imap for account that has huge number of emails in inbox but it is working for others accounts that doesn't have that many emails. May be it is not the amount of imap commands but overral traffic that particular imap command generate. Anyway I just gave up chasing this issue and trying to figure out what the hell is going on with microsoft servers. I advice you not to waste you time with those servers. I think some vendors (like MS and Yahoo) intentionally sabotage IMAP to ruin users experience with desktop email client and force you to use their Web UI so they can show you ads or profit by some other way from your visits.
